File: sorting/MergeSort.java
/*
Problem: Merge Sort
Description: Implement merge sort algorithm to sort an array in ascending order.

Time Complexity: O(n log n)
Space Complexity: O(n)

Approach:
- Use recursive divide and conquer.
- Split array into halves, recursively sort them, then merge back.
*/

import java.util.Arrays;

public class MergeSort {

    public static void mergeSort(int[] arr, int left, int right) {
        if (left < right) {
            int mid = (left + right) / 2;

            // Sort first and second halves
            mergeSort(arr, left, mid);
            mergeSort(arr, mid + 1, right);

            // Merge sorted halves
            merge(arr, left, mid, right);
        }
    }

    private static void merge(int[] arr, int left, int mid, int right) {
        int n1 = mid - left + 1;
        int n2 = right - mid;

        int[] L = new int[n1];
        int[] R = new int[n2];

        // Copy data into temporary arrays
        for (int i = 0; i < n1; i++) L[i] = arr[left + i];
        for (int j = 0; j < n2; j++) R[j] = arr[mid + 1 + j];

        int i = 0, j = 0, k = left;

        // Merge two halves
        while (i < n1 && j < n2) {
            if (L[i] <= R[j]) arr[k++] = L[i++];
            else arr[k++] = R[j++];
        }

        // Copy remaining elements
        while (i < n1) arr[k++] = L[i++];
        while (j < n2) arr[k++] = R[j++];
    }

    public static void main(String[] args) {
        int[] arr = {5, 2, 4, 6, 1, 3};
        System.out.println("Original array: " + Arrays.toString(arr));

        mergeSort(arr, 0, arr.length - 1);

        System.out.println("Sorted array: " + Arrays.toString(arr));
    }
}

File: sorting/merge-sort.py
"""
Problem: Merge Sort
Description: Implement merge sort algorithm to sort an array in ascending order.

Time Complexity: O(n log n)
Space Complexity: O(n)

Approach:
- Use the recursive divide-and-conquer strategy.
- Divide the array into halves until single elements remain.
- Merge the sorted halves into a fully sorted array.
"""

def merge_sort(arr):
    # Base case: if array has 1 or fewer elements, it's already sorted
    if len(arr) <= 1:
        return arr

    mid = len(arr) // 2

    # Recursively sort both halves
    left_half = merge_sort(arr[:mid])
    right_half = merge_sort(arr[mid:])

    # Merge the sorted halves
    return merge(left_half, right_half)


def merge(left, right):
    merged = []
    i = j = 0

    # Merge while elements remain in both arrays
    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            merged.append(left[i])
            i += 1
        else:
            merged.append(right[j])
            j += 1

    # Append remaining elements
    merged.extend(left[i:])
    merged.extend(right[j:])

    return merged


# Test cases
if __name__ == "__main__":
    arr = [5, 2, 4, 6, 1, 3]
    print("Original array:", arr)
    print("Sorted array:", merge_sort(arr))
